source/extensions/common/aws/credentials_provider_impl.cc

```diff
@@ -253,6 +335,9 @@ DefaultCredentialsProviderChain::DefaultCredentialsProviderChain(
   ENVOY_LOG(debug, "Using environment credentials provider");
   add(factories.createEnvironmentCredentialsProvider());
 
+  ENVOY_LOG(debug, "Using credentials file credentials provider");
+  add(factories.createCredentialsFileCredentialsProvider(api));
```

Comment:
I am contemplating whether we want to include this credentials provider by default.

Because if the file is not present then `CredentialsFileCredentialsProvider::refresh()` is called again and again because last_updated_ is not updated in failure scenario.


Prefer if don't include it by default so that it won't affect other providers. Include only if `~/.aws/credentials` or AWS_SHARED_CREDENTIALS_FILE is present. This means after Envoy has initialized we will not be able to updated the credentials in the background without restarting Envoy.

Reply:
welcome to any other suggestions so that we don't affect performance of other credentials provider.

Reply:
That's a good point, thanks for raising it. These are the main options I can think of:
1. Add a field to the AwsRequestSigning proto to enable the credentials file. If it's enabled, we add it to the default chain.
2. Add CredentialsFileCredentialsProvider to the default chain iff the credentials file exists.
3. Add CredentialsFileCredentialsProvider to the default chain, but have refreshIfNeeded return immediately unless AWS_SHARED_CREDENTIALS_FILE and AWS_PROFILE are both defined.

I prefer option 1 because it won't impact any existing users of the filter, and since fetching from the credentials file can be expensive (especially if credentials aren't found) it requires the user to explicitly enable it. We can document that it's expensive and should only be enabled if the user is sure credentials will be found in it. What do you think?

Reply:
I prefer option 2 because that is the approach with least intrusion. Today for other credentials provider we don't expect any configuration on AWS Lambda or AWS Request signing filter for deciding on the credentials provider. It is inline with current implementation of tasks credentials provider where it expects a particular ENV variable to be available. In credentials file credentials provider we expect a file to be present on default path or on the path specified by `AWS_SHARED_CREDENTIALS_FILE`. This is checked only once at the filter initialization.

This comes with a drawback that the file has to be present before Envoy consumes the configuration.

For host & task credentials provider no impact if the file is not present. But if the file is present (for some other purpose) and users don't expect the Envoy to pick the credentials specified on the file then we can add a runtime config to disable this feature.

Reply:
Another drawback is that if the credentials file is ever deleted for some reason then the filter will start trying to read the file on every request, though I imagine that's unlikely.

The change you suggested is reasonable and sounds good to me, but let me float one more idea: if `CredentialsFileCredentialsProvider::extractCredentials` returns early, we cache an empty set of credentials and update `last_updated_`. That way if we fail to fetch credentials for any reason, we will cache empty credentials for an hour and the next provider in the chain will be called. This won't be that impactful if the credentials file doesn't exist since we will only attempt to read the file once an hour. The benefits are that the file doesn't have to be present before Envoy consumes the config, it accommodates the credentials file being deleted, and it accommodates the value of `AWS_SHARED_CREDENTIALS_FILE` changing.

Reply:
This is a good option as well. Practically it could be the same as option 2 because there is no case where Envoy would be considered operational given it went 1 hour without any credentials. Make sure we add debug logs around credentials file not found so that user won't be confused what's happening under the hood.

Would you still be willing to add a runtime configuration so that we have an option to disable credential file credentials provider if we have to? Especially for those use case where it already has credentials file in the Envoy container path but instead prefer to use Instance metadata or task metadata. After 1 year time period (deprecation of runtime configs) when we decide to remove the runtime configuration we would have registered if Envoy users needed to use the runtime config then we can provide a permanent xDS config option to disable the credentials file credentials provider.

Reply:
Yep, I will add the runtime config to both the Request Signing and Lambda filters per your suggestion.

Reply:
This is done.